File: src/leetcode/2026/01/Solution530.cpp
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution530 {
public:
    int getMinimumDifference(TreeNode* root) {
        int minDiff = INT_MAX;
        TreeNode* prev = nullptr;
        inorderTraversal(root, prev, minDiff);
        return minDiff;
    }

private:
    void inorderTraversal(TreeNode* node, TreeNode*& prev, int& minDiff) {
        if (!node) return;

        inorderTraversal(node->left, prev, minDiff);

        if (prev) {
            minDiff = min(minDiff, node->val - prev->val);
        }
        prev = node;

        inorderTraversal(node->right, prev, minDiff);
    }
};

int main() {
    Solution530 solution;

    // Example usage:
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(2);
    root->right = new TreeNode(6);
    root->left->left = new TreeNode(1);
    root->left->right = new TreeNode(3);

    cout << solution.getMinimumDifference(root) << endl; // Expected output: 1

    // Clean up memory (not shown here for brevity)
    delete root->left->left;
    delete root->left->right;
    delete root->left;
    delete root->right;
    delete root;

    return 0;
}

File: src/leetcode/2026/01/Solution530.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution530 {
    public int getMinimumDifference(TreeNode root) {
        int[] prev = { -1 };
        int[] minDiff = { Integer.MAX_VALUE };
        inorderTraversal(root, prev, minDiff);
        return minDiff[0];
    }

    private void inorderTraversal(TreeNode node, int[] prev, int[] minDiff) {
        if (node == null) {
            return;
        }

        inorderTraversal(node.left, prev, minDiff);

        if (prev[0] != -1) {
            minDiff[0] = Math.min(minDiff[0], node.val - prev[0]);
        }
        prev[0] = node.val;

        inorderTraversal(node.right, prev, minDiff);
    }

    public static void main(String[] args) {
        Solution530 solution = new Solution530();

        TreeNode root1 = new TreeNode(4,
                new TreeNode(2,
                        new TreeNode(1),
                        new TreeNode(3)),
                new TreeNode(6));
        System.out.println("The minimum absolute difference in the BST is: " +
                solution.getMinimumDifference(root1));

        TreeNode root2 = new TreeNode(1,
                new TreeNode(0),
                new TreeNode(48,
                        new TreeNode(12),
                        new TreeNode(49)));
        System.out.println("The minimum absolute difference in the BST is: " +
                solution.getMinimumDifference(root2));
    }
}
